_inc/lib/class.core-rest-api-endpoints.php

```diff
@@ -415,6 +415,35 @@ public static function register_endpoints() {
 				),
 			)
 		) );
+
+		// Get and set API keys.
+		register_rest_route(
+			'jetpack/v4',
+			'/api-key/(?P<service>[a-z\-_]+)',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => __CLASS__ . '::get_api_key',
+					'permission_callback' => __CLASS__ . '::view_admin_page_permission_check',
```

Comment:
Echoing @enejb's [comment](https://github.com/Automattic/jetpack/commit/929cbe3ee6720fddf0704818a877779cd2de0784#r31027363), I think we need to restrict this more.

`::view_admin_page_permission_check()` just wraps a check for the `jetpack_admin_page` capability, which is just [`read`](https://github.com/Automattic/jetpack/blob/a5cb74e97db5a1df5134827eb5d28b3fb7051c87/class.jetpack.php#L917-L924).

Reply:
@mdawaffe, @enejb what would your suggestions for `permission_callback` be? For the GET endpoint we need it to be entirely unrestricted, since this is accessed from the frontend by non-editors. For POST and DELETE, I'm less sure what the appropriate restriction would be?

Reply:
> What's the frontend?

Sorry, I meant "any old visitors to the website". Is there a more accepted way to describe?

`:: update_settings_permission_check()` sounds good. My only question is whether there's a category of user who can see & edit the block, but cannot set or change the API key? This would be a bad situation, since the block is a brick without the key. So, if an editor adds the block for the first time but is blocked from setting the Google API key, they are stuck.

Reply:
I'm hoping to avoid getting into the business of assessing the current user's access privileges unless absolutely necessary—I haven't explored it yet but it may be a little hairy to implement in these early wild west days of Gutenberg, plus it adds a new layer of testing requirements. Scenarios 1, 2, and 4 all require this level of sophistication. And I'm also hoping to avoid per-block keys (scenario 3), because that will get mega-annoying for the user if they use maps with any regularity. 

One major advantage of a site-wide API key is that it can be globally swapped if compromised, if a new entity takes over the billing, etc. Imagine a nasty little situation where a site has hundreds of maps and has to edit every one of them if for some reason their key needs to be updated. 

My $0.02: What's the argument against simply letting Authors add/update the key? It seems to me that on balance, the risk of malicious intent there is more palatable than complicating the UI for setting up the key. Also note that this key is currently used for nothing but the Map block, so if it gets inadvertently messed up, the risk isn't all that dire. If you feel differently, though, I'm happy to dig in! 

Also, another note: my eventual hope is that in future iterations of the API, Automattic provides a centralized key which can be used only in the editor and preview. This will allow any JP customer to use the block immediately—they will only have to add a key when they are ready to begin publishing. When this level of sophistication becomes reality might be a good time to tighten up the permissions, since we will no longer be as concerned about the risks of a bad first experience.

Reply:
As it stands, anyone with an account (even people with the "Subscriber" role) can add/edit/delete the key. Imagine a large site that's set up to encourage frequent commenters to sign up for the subscriber role (see wp-admin -> Settings -> General -> "Anyone can register").

Even for Authors (i.e., users with the "Author" role), this still seems odd; in general, Authors in WordPress are allowed to do things to their posts, but are not allowed to influence others' posts or site-wide settings. Suppose an Author, before leaving an organization's site, updates the key to one they control. Now an ex-employee (or whatever) has access to stats of the site and can even break the site by invalidating the key.

Yes, this sort of thing would still be a problem with an Admin or Editor leaving, but organizations know to audit things when users of that role leave. Since WordPress is set up to limit the access Authors have, though, most organizations wouldn't think much about Authors leaving.

This is all related to the block visibility discussion at p1HpG7-5JG-p2.

Perhaps a compromise?:
* Subscribers and Contributors can not write to this API at all
* Authors and Editors can add a key but not edit or delete one
* Admins can add, edit, or delete.

(Or perhaps group Editors and Admins together?)

Also, I believe it's the case that the vast majority of sites only have one user account. I don't know if the vast majority of "interesting" (higher traffic, paid customers, publishing volume, maps block users, whatever metric you pick) sites have only one user account, though.

Reply:
Thinking about this a little more I don't think there's a meaningful difference between CREATE and UPDATE here since the API keys are stored as site options with only one value per site. There isn't really a difference between unset and blank. So, instead of separating the two actions I adopted your stricter suggestion for UPDATE/DELETE (editors/admins only), and removed all restrictions on GET. I'll build some checking/messaging into the block so that < Editors understand why they are unable to set the key. Changes are in https://github.com/Automattic/jetpack/pull/10393/commits/516c7878dcb8dbbdfe0972a1f5c735abdff8331b.